model/src/application/interaction/application_command/data/mod.rs

```diff
@@ -26,50 +30,168 @@ pub struct CommandData {
 
 /// Data received when a user fills in a command option.
 ///
-/// Note: user, channel, role, and mentionable option types will be returned as
-/// a [`String`] option here.
-///
 /// Refer to [the discord docs] for more information.
 ///
 /// [the discord docs]: https://discord.com/developers/docs/interactions/slash-commands#interaction-applicationcommandinteractiondataoption
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
-#[serde(untagged)]
-pub enum CommandDataOption {
-    String {
-        name: String,
-        value: String,
-    },
-    Integer {
-        name: String,
-        value: i64,
-    },
-    Boolean {
-        name: String,
-        value: bool,
-    },
-    SubCommand {
-        name: String,
-        #[serde(default)]
-        options: Vec<CommandDataOption>,
-    },
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct CommandDataOption {
+    pub name: String,
+    pub value: CommandOptionValue,
 }
 
-impl CommandDataOption {
-    pub const fn kind(&self) -> &'static str {
-        match self {
-            CommandDataOption::String { .. } => "String",
-            CommandDataOption::Integer { .. } => "Integer",
-            CommandDataOption::Boolean { .. } => "Boolean",
-            CommandDataOption::SubCommand { .. } => "SubCommand",
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum CommandOptionValue {
+    String(String),
+    Integer(i64),
+    Boolean(bool),
+    User(UserId),
+    Channel(ChannelId),
+    Role(RoleId),
+    Mentionable(u64),
+    SubCommand(Vec<CommandDataOption>),
+    SubCommandGroup(Vec<CommandDataOption>),
+}
+
+#[derive(Debug, Deserialize)]
+struct CommandDataOptionRaw<'a> {
+    name: String,
+    #[serde(rename = "type")]
+    kind: CommandOptionType,
+    value: Option<CommandOptionValueRaw<'a>>,
+    #[serde(default)]
+    options: Option<Vec<CommandDataOption>>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+enum CommandOptionValueRaw<'a> {
+    String(Cow<'a, str>),
+    Integer(i64),
+    Boolean(bool),
+}
+
+impl Serialize for CommandDataOption {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut state = serializer.serialize_struct("CommandDataOptionRaw", 3)?;
+        state.serialize_field("name", &self.name)?;
+        state.serialize_field("type", &self.value.kind())?;
+        match &self.value {
+            CommandOptionValue::SubCommand(opts) | CommandOptionValue::SubCommandGroup(opts) => {
+                state.serialize_field("options", &Some(opts))?
+            }
+            _ => state.serialize_field("value", &Some(self.value.raw_value().unwrap()))?,
```

Comment:
There's an unwrap here

Reply:
This is guaranteed not to panic because we handle `SubCommand` and `SubCommandGroup` above. Maybe I should just inline `raw_value` here.

Reply:
Indeed, inlining sounds like a better option here